File: examples/named-pipes/nodejs/can_send.js
#!/usr/bin/env node
// Transmit a CAN frame through a rexgend named pipe.
//   usage: node can_send.js [bus] [ID_hex] [byte ...]
//     e.g. node can_send.js can0 153 AA BB CC DD EE FF 00 11
//          node can_send.js can0 1ABCDEF0 DE AD BE EF        # extended ID
//
// TX line format:  <ts> canN [flags] <ID_hex> <dlc> b0 b1 ...
//   <ts> ignored by the daemon -> use 0
//   ID_hex 1-3 hex = standard, 4+ = extended

const fs = require('fs');

const bus = process.argv[2] || 'can0';
const canId = process.argv[3] || '153';
const data = process.argv.length > 4
  ? process.argv.slice(4)
  : ['AA', 'BB', 'CC', 'DD', 'EE', 'FF', '00', '11'];

const path = `/var/run/rexgen/${bus}/tx`;
// add "[FB] " before canId for CAN-FD+BRS
const line = `0 ${bus} ${canId} ${data.length} ${data.join(' ')}`;

try {
  fs.writeFileSync(path, line + '\n');              // one frame per write
  console.log(`sent on ${bus}: ${line}`);
} catch (e) {
  console.error(`cannot write ${path} (rexgend running / ${bus} tx configured?) - ${e.code}`);
  process.exit(1);
}

File: examples/named-pipes/python/can_send.py
#!/usr/bin/env python3
"""Transmit a CAN frame through a rexgend named pipe.

usage: ./can_send.py [bus] [ID_hex] [byte ...]
   e.g. ./can_send.py can0 153 AA BB CC DD EE FF 00 11
        ./can_send.py can0 1ABCDEF0 DE AD BE EF          # extended ID

TX line format:  <ts> canN [flags] <ID_hex> <dlc> b0 b1 ...
  <ts>     ignored by the daemon (it stamps the frame) -> use 0
  ID_hex   1-3 hex digits = standard, 4+ = extended
"""
import sys

bus = sys.argv[1] if len(sys.argv) > 1 else "can0"
can_id = sys.argv[2] if len(sys.argv) > 2 else "153"
data = sys.argv[3:] if len(sys.argv) > 3 else ["AA", "BB", "CC", "DD", "EE", "FF", "00", "11"]

path = f"/var/run/rexgen/{bus}/tx"
# <ts> <bus> <id> <dlc> <bytes...>   (add "[FB]" before <id> for CAN-FD+BRS)
line = f"0 {bus} {can_id} {len(data)} " + " ".join(data)

try:
    with open(path, "w") as fp:
        fp.write(line + "\n")
    print(f"sent on {bus}: {line}")
except FileNotFoundError:
    sys.exit(f"no pipe {path} (is rexgend running / {bus} tx configured?)")
